### ISSUE TITLE
`outdated` & `upgrade` fail on Catalina

### ISSUE THREAD

**Issue by revolter** [mas-cli/mas#111]
Running on macOS High Siera Version 10.13.1 (17B1003), `mas outdated` doesn't print anything, even if there is an update for macOS 10.13.2 and I have the `App Store` opened.

**Comment by jitendravyas** [comment]
Same here

**Comment by wittyfans** [comment]
10.13.2, Input mas signout does not work, Can't sign out my account.

**Comment by fbartho** [comment]
`mas outdated` also doesn't work on 10.14 Mojave

**Comment by rasmusbe** [comment]
![image](https://user-images.githubusercontent.com/2124386/48199355-1e523300-e35c-11e8-9dab-2a146cbbf893.png)

![image](https://user-images.githubusercontent.com/2124386/48199404-37f37a80-e35c-11e8-939c-eb73118029a8.png)

**Comment by phatblat** [comment]
@rasmusbe what macOS version are you using?

**Comment by phatblat** [comment]
Tested this today using mas 1.4.3 on macOS 10.14.1 and it appears to be working.

```
$ mas outdated
406056744 Evernote (7.5.2 -> 7.6)
```

![screen shot 2018-11-16 at 12 41 09 pm](https://user-images.githubusercontent.com/28851/48643765-d5bef900-e99d-11e8-9a7b-885b2da4373b.png)

Wondering if this spontaneous fix is related to #182

**Comment by Henrietta1989** [comment]
Numbers 5.3 is latest and outdated don’t find it in 10.14.1:

`Home:~ Henrietta$ mas list
497799835 Xcode (10.1)
409203825 Numbers (5.2)
Home:~ Henrietta $ mas outdated
Home:~ Henrietta $ `

**Comment by pavstor** [comment]
Have the same issue on Mojave 10.14.1:

<img width="1250" alt="screen shot 2018-11-17 at 10 59 47" src="https://user-images.githubusercontent.com/38346023/48659877-69550000-ea58-11e8-99b8-e2b00ccbf415.png">

**Comment by drod3763** [comment]
No its not fixed. It only detects outdated programs for me _after_ i've already seen them in the Mac App Store.

**Comment by rasmusbe** [comment]
@phatblat I was using mas 1.4.3 on macOS 10.14.1

**Comment by spinofdoom** [comment]
Just piping in to say that 'outdated' and 'upgrade' are still broken for me in Mojave 10.14.1.

```
mas list

425264550 Blackmagic Disk Speed Test (3.2)
937984704 Amphetamine (4.0.4)
409201541 Pages (7.3)
1176895641 Spark (2.1.2)

mas search amphetamine

937984704  Amphetamine (4.0.5)

mas outdated

mas upgrade
Everything is up-to-date
```

**Comment by phatblat** [comment]
@spinofdoom thanks for confirming. As a test, can you open the MAS GUI app, loop at the pending updates and then rerun those commands afterwards?

**Comment by rasmusbe** [comment]
@phatblat today when I woke my computer up I got the MAS notification about available updates and decided to do some testing.

1. Ran `mas outdated` directly, no updates
2. Opened MAS GUI, one available update there
3. Ran `mas outdated` directly, no updates
4. Waited a minute, ran `mas outdated` again - one update!

This was done using mas cli 1.4.3 on macOS 10.14.1.

**Comment by rasmusbe** [comment]
@revolter it's the gui app built by Apple :)

**Comment by spinofdoom** [comment]
@phatblat Sorry, I already went ahead and updated via App Store. I'll keep y'all posted next time updates come up.

**Comment by phatblat** [comment]
I missed that #76 already captures the issue. `mas outdated` doesn't trigger a check for updates, only displays _known_ pending updates. This is a limitation of the tool as it stands today. Hopefully, we can find an API to trigger the check.

Closing this issue in favor of #76

**Comment by rasmusbe** [comment]
Oh, I see. But I do get an notification about available updates before `mas outdated` so the system must know about them somehow. I just hope we can find a way to access it.

And just to complete the debugging, `mas reset` does not help. 
<img width="351" alt="image" src="https://user-images.githubusercontent.com/2124386/49067599-ce60e080-f224-11e8-8906-edbfbc15f51e.png">
![image](https://user-images.githubusercontent.com/2124386/49067760-37e0ef00-f225-11e8-93dd-a755fbff29f4.png)
![image](https://user-images.githubusercontent.com/2124386/49067681-02d49c80-f225-11e8-99c2-736325e37800.png)

**Comment by phatblat** [comment]
@revolter thanks for the clarification. I will reopen, but this is a tricky one to test as it can only be reproduced by installing an app, that app being updated and then testing.

**Comment by spinofdoom** [comment]
Finally got an update to test.

With App Store open and update showing in Mojave (10.14.2), using mas 1.4.3:

```
$ mas list
425264550 Blackmagic Disk Speed Test (3.2)
937984704 Amphetamine (4.0.5)
1176895641 Spark (2.1.2)
409201541 Pages (7.3)
```

```
$ mas search spark
  1176895641  Spark - Email App by Readdle              (2.1.3)
   ...
```
```
$ mas outdated

```
```
$ mas upgrade
Everything is up-to-date
```


Identical after `mas reset` and even logging out and back in to the App Store GUI.

**Comment by revolter** [comment]
Where is the code that deals with the list of apps that need updating?

**Comment by phatblat** [comment]
It's in [`OutdatedCommand`](https://github.com/mas-cli/mas/blob/master/MasKit/Commands/Outdated.swift#L30-L37), which relies on the private [`-[CKUpdateController availableUpdates]`](https://github.com/mas-cli/mas/blob/master/PrivateFrameworks/CommerceKit/CKUpdateController.h#L93).

**Comment by apizz** [comment]
Can confirm this is still in an issue in latest 1.6.3 release.

`mas outdated` does not reveal any updates.  However, upon opening Mac App Store app and clicking 'Updates' see two available app updates.  Subsequent `mas outdated` reflects this.

**Comment by wkuckro** [comment]
ditto to @apizz. `mas outdated` returned no results, `mas list` returns ```
682658836 com.apple.garageband10 (10.3.2)
408981434 com.apple.iMovieApp (10.1.10)
497799835 com.apple.dt.Xcode (10.1)
425264550 com.blackmagic-design.DiskSpeedTest (3.2)
409183694 com.apple.iWork.Keynote (8.3)
409201541 com.apple.iWork.Pages (7.3)
409203825 com.apple.iWork.Numbers (5.3)```

And the Mac App store app has updates for Xcode -> 10.2, Pages -> 8.0, iMovie -> 10.1.11, Numbers  -> 6.0, and Keynote -> 9.01

**Comment by moisiss** [comment]
I can confirm that this is still an issue...  mas list returns Keynote (9.0.1) (among other apps that I have installed).  The GUI app store says that Keynote 9.0.2 is a pending update, but when I run mas outdated nothing is returned and when I run mas upgrade it tells me that everything is up to date.  Any work around or fix for this yet?

**Comment by tomosuto** [comment]
App Store GUI was showing there were 2 pending apps.
So I did `mas outdated` but got only one app. 

```
$ mas outdated  
410628904 Wunderlist: To-Do List & Tasks (3.4.10 -> 3.4.11)
```

```
$ mas upgrade
Upgrading 1 outdated application:
Wunderlist: To-Do List & Tasks (3.4.11)
```

```
$ mas upgrade
Everything is up-to-date
```

Another app that should be returned is Keynote (9.0.1 -> 9.0.2).

![Screen Shot 2019-05-09 at 2 40 03 AM](https://user-images.githubusercontent.com/15020398/57432401-c8275500-7203-11e9-83cc-a4b9894cc918.png)

**Comment by tomosuto** [comment]
A day after I posted the post above, I did `mas outdated` again and it showed Keynote upgrade.

```
$ mas outdated
409183694 Keynote (9.0.1 -> 9.0.2)
```
```
$ mas upgrade
Upgrading 1 outdated application:
Keynote (9.0.2)
==> Downloading Keynote
==> Installed Keynote
```
```
$ mas upgrade
Everything is up-to-date
```
@moisiss Probably It will now work for you too.

**Comment by ruchern** [comment]
Possibly related to #244 on macOS Catalina. Will monitor this issue too.

**Comment by melgu** [comment]
Still doesn't work (on Catalina):
```
> mas list | rg Highland
1171820258 Highland 2 (2.6.7)
>>> mas search Highland
  1171820258  Highland 2 (2.6.8)
>>> mas outdated
>>> mas upgrade 1171820258
Warning: Nothing found to upgrade
```

I even have the GUI App Store Update page open.

**Comment by newz2000** [comment]
As a work-around, can the `mas upgrade` command simply try harder to upgrade? For example, I know my Final Cut Pro is outdated, so I tried this:
    ~ mas install 424389933
    Warning: Final Cut Pro is already installed
    ~ mas upgrade 424389933
    Warning: Nothing found to upgrade

Yet there is an update available:
    ~ mas list
    424389933 Final Cut Pro (10.4.6)
    ~ mas info 424389933
    Final Cut Pro 10.4.7 [299.99]

I would be OK if `mas upgrade` would check with Apple for the update, or if `mas install` would do a reinstall of the app if it's already installed.

**Comment by tabchas** [comment]
Still looks like it's broken. There are updates that show up in the GUI but won't show up in the the `mas outdated` list. It also is strangely showing that Xcode needs to be downgraded from 11.3.1 to 11?

![Screen Shot 2020-02-22 at 10 14 32 AM](https://user-images.githubusercontent.com/732771/75095629-82eace80-555c-11ea-8d67-e421e99edff7.png)

**Comment by rogeriopradoj** [comment]
> As a work-around, can the `mas upgrade` command simply try harder to upgrade? For example, I know my Final Cut Pro is outdated, so I tried this:
> ~ mas install 424389933
> Warning: Final Cut Pro is already installed
> ~ mas upgrade 424389933
> Warning: Nothing found to upgrade
> 
> Yet there is an update available:
> ~ mas list
> 424389933 Final Cut Pro (10.4.6)
> ~ mas info 424389933
> Final Cut Pro 10.4.7 [299.99]
> 
> I would be OK if `mas upgrade` would check with Apple for the update, or if `mas install` would do a reinstall of the app if it's already installed.

Pointing to @newz2000 said, I had to force install, and it did the job via cli.

But, for that, I had to do these steps:

- `mac list` to get all the app ids
- open the App Store.app to see which updates are available (running `mac outdated` shows nothing)
- run `mac install ID_OF_APP --force` for each of the outdated apps

After that, I got it all upgraded.

I think it is much more painful than just go to App Store.app and click in **Update All** unfortunately

**Comment by jimothyGator** [comment]
@rogeriopradoj 's comment made me think of an idea: Could `outdated` run `info` for each installed app, and compare either the version number or release date? Obviously, the more apps you have installed, the slower this would be. 

Maybe keep the existing behavior (using `CKUpdateController`), but with `outdated --force` and `upgrade --force`, use this slower method.

**Comment by phatblat** [comment]
Fixed now in [1.8.1](https://github.com/mas-cli/mas/releases/tag/v1.8.1)
